[SPARK-11249][launcher] Throw error if app resource is not provided.

vanzin opened the pull request:
Without this, the code would build an invalid spark-submit command line,
and a more cryptic error would be presented to the user. Also, expose
a constant that allows users to set a dummy resource in cases where
they don't need an actual resource file; for backwards compatibility,
that uses the same "spark-internal" resource that Spark itself uses.

Tested via unit tests, run-example, spark-shell, and running the
thrift server with mixed spark and hive command line arguments.

Commit:
[SPARK-11249][launcher] Throw error is app resource is not provided.

Without this, the code would build an invalid spark-submit command line,
and a more cryptic error would be presented to the user. Also, expose
a constant that allows users to set a dummy resource in cases where
they don't need an actual resource file; for backwards compatibility,
that uses the same "spark-internal" resource that Spark itself uses.

Tested via unit tests, run-example, spark-shell, and running the
thrift server with mixed spark and hive command line arguments.

SparkQA commented:
**[Test build #57807 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57807/consoleFull)** for PR 12909 at commit [`0ab4a28`](https://github.com/apache/spark/commit/0ab4a2864f17189898376ff4f7e02e8b7b2c416e).

SparkQA commented:
**[Test build #57807 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57807/consoleFull)** for PR 12909 at commit [`0ab4a28`](https://github.com/apache/spark/commit/0ab4a2864f17189898376ff4f7e02e8b7b2c416e).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57807/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #57900 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57900/consoleFull)** for PR 12909 at commit [`0ab4a28`](https://github.com/apache/spark/commit/0ab4a2864f17189898376ff4f7e02e8b7b2c416e).

SparkQA commented:
**[Test build #57900 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57900/consoleFull)** for PR 12909 at commit [`0ab4a28`](https://github.com/apache/spark/commit/0ab4a2864f17189898376ff4f7e02e8b7b2c416e).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57900/
Test FAILed.

vanzin commented:
Hmm, pyspark seems to be getting stuck, let me see if I broke it...

Commit:
Fix python / R shells.

Because the parser now sets appResource instead of adding it to
the raw args list, the code that creates the command line for those
shells need to override it with the right value before setting the
final command line.

SparkQA commented:
**[Test build #58030 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58030/consoleFull)** for PR 12909 at commit [`46664ba`](https://github.com/apache/spark/commit/46664bad2b0398260771a88a3f206cf1e1da0750).

SparkQA commented:
**[Test build #58030 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58030/consoleFull)** for PR 12909 at commit [`46664ba`](https://github.com/apache/spark/commit/46664bad2b0398260771a88a3f206cf1e1da0750).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58030/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #58039 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58039/consoleFull)** for PR 12909 at commit [`46664ba`](https://github.com/apache/spark/commit/46664bad2b0398260771a88a3f206cf1e1da0750).

SparkQA commented:
**[Test build #58039 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58039/consoleFull)** for PR 12909 at commit [`46664ba`](https://github.com/apache/spark/commit/46664bad2b0398260771a88a3f206cf1e1da0750).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58039/
Test PASSed.

vanzin commented:
@andrewor14

vanzin commented:
Merging to master / 2.0.
